### ISSUE TITLE
ECS Agent randomly fails - shows as "Agent Connected: false" in ECS Cluster

### ISSUE THREAD

**Issue by schuylr** [rancher/os#671]
Sporadically, we will get rancheros instances that no longer have a working ECS Agent. When I look at `/var/log/ecs` on the ros host, I find a bunch of files with this message repeated:

>2015-12-01T09:59:59Z [ERROR] Unable to connect to docker daemon . Ensure docker is running module="TaskEngine" endpoint="unix:///var/run/docker.sock" err="dial unix /var/run/docker.sock: connection refused"

Yet I can still connect to docker:

```
[rancher@ip-10-233-98-202 run]$ docker info
Containers: 2
Images: 53
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.3-rancher
Operating System: RancherOS (containerized)
CPUs: 2
Total Memory: 7.307 GiB
```

Restarting the instance doesn't fix the issue, nor does restarting the `amazon-ecs-agent` using `sudo ros service restart amazon-ecs-agent` or `ros kill`.

Our only recourse for now is to kill the instance to be replaced automatically by the auto scaler.

**Comment by protomouse** [comment]
Encountering this as well. Did you learn anything?

**Comment by schuylr** [comment]
Unfortunately no. I don't know what other information I can post here because I don't know what else to look for!

**Comment by wonderslug** [comment]
Im getting this as well.  I seem to encounter it after any reboot of the rancheros instance.  

Of note I have noticed that prior to the reboot the ecs-agents name in the system docker is os_ecs-agent_1 and after the reboot its ecs-agent.

**Comment by wonderslug** [comment]
Did a diff of an inspect of a working and a non-working ecs instance and the only thing that looks different are ids.

Attached are my inspects   i did reorder the volumes and mounts to make them line up for any diffs but there were not any.

[broken-ecs_inspect.txt](https://github.com/rancher/os/files/110488/broken-ecs_inspect.txt)
[working_ecs_inspect.txt](https://github.com/rancher/os/files/110489/working_ecs_inspect.txt)

**Comment by madj4ck** [comment]
Face the same issue too. What I did was to remove the os_ecs-agent_1 container using 'system-docker rm os_ecs-agent_1' and recreate the agent container via instruction on http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-agent-install.html.

**Comment by schuylr** [comment]
This is still an issue on v0.4.4, now more than ever. I can't get it to connect to ECS even somewhat reliably and the error logs still say the same thing as noted in the first comment.

**Comment by ibuildthecloud** [comment]
I believe the issue here is that the ECS agent is bind mounting `/var/run/docker.sock` and not `/var/run`.  Looking at the below  

```yaml
ecs-agent:
  image: amazon/amazon-ecs-agent
  labels:
    io.rancher.os.scope: system
    io.rancher.os.after: docker
  restart: always
  volumes:
  - /var/run/docker.sock:/var/run/docker.sock
  - /var/log/ecs/:/log
  - /var/lib/ecs/data:/data
  - /sys/fs/cgroup:/sys/fs/cgroup:ro
  - /var/run/docker/execdriver/native:/var/lib/docker/execdriver/native:ro
  ports:
  - "127.0.0.1:51678:51678"
  environment:
  - ECS_LOGFILE=/log/ecs-agent.log
  - ECS_LOGLEVEL=info
  - ECS_DATADIR=/data
  - ECS_*
  - AWS_*
```

**Comment by schuylr** [comment]
@ibuildthecloud Do you suppose the issue stems from `/var/run/docker.sock` not existing on boot when the `system-docker` and `ecs-agent` are subsequently started? Then the system docker would initialize `docker.sock` as a mounted folder instead of the file directly? That would explain why manually killing and restarting it would work.

So changing the `ecs-agent` config in the image to mount `/var/run` instead of `/var/run/docker.sock` could solve the problem, yes?

**Comment by schuylr** [comment]
I can confirm that by running my own service called `ecs-agent-custom` and using a non-ECS RancherOS AMI v0.4.4, I get a successful connection every time! See logs below. Note the repeated failures until the user docker initializes.

```
2016-04-19T14:04:04Z [INFO] Starting Agent: Amazon ECS Agent - v1.8.2 (64205ac)
2016-04-19T14:04:04Z [INFO] Loading configuration
2016-04-19T14:04:04Z [WARN] Invalid value for task cleanup duration, will be overridden to 3h0m0s module="config" parsed value="0" minimum threshold="1m0s"
2016-04-19T14:04:04Z [INFO] Checkpointing is enabled. Attempting to load state
2016-04-19T14:04:04Z [INFO] Loading state! module="statemanager"
2016-04-19T14:04:04Z [ERROR] Unable to connect to docker daemon . Ensure docker is running module="TaskEngine" endpoint="unix:///var/run/docker.sock" err="dial unix /var/run/docker.sock: connection refused"
2016-04-19T14:04:04Z [INFO] Registering Instance with ECS
2016-04-19T14:04:05Z [INFO] Registered! module="api client"
2016-04-19T14:04:05Z [INFO] Registration completed successfully. I am running as 'arn:aws:ecs:us-east-1:982267928111:container-instance/0e635bc9-2206-48fb-9b76-53b3e6a6f71b' in cluster 'jewlr-us'
2016-04-19T14:04:05Z [INFO] Saving state! module="statemanager"
2016-04-19T14:04:05Z [ERROR] Unable to connect to docker daemon . Ensure docker is running module="TaskEngine" endpoint="unix:///var/run/docker.sock" err="dial unix /var/run/docker.sock: connection refused"
2016-04-19T14:04:05Z [ERROR] Could not connect to docker daemon module="TaskEngine" err="dial unix /var/run/docker.sock: connection refused"
2016-04-19T14:04:05Z [ERROR] Unable to connect to docker daemon . Ensure docker is running module="TaskEngine" endpoint="unix:///var/run/docker.sock" err="dial unix /var/run/docker.sock: connection refused"
2016-04-19T14:04:05Z [ERROR] Unable to connect to docker daemon . Ensure docker is running module="TaskEngine" endpoint="unix:///var/run/docker.sock" err="dial unix /var/run/docker.sock: connection refused"
2016-04-19T14:04:06Z [ERROR] Unable to connect to docker daemon . Ensure docker is running module="TaskEngine" endpoint="unix:///var/run/docker.sock" err="dial unix /var/run/docker.sock: connection refused"
2016-04-19T14:04:06Z [ERROR] Unable to connect to docker daemon . Ensure docker is running module="TaskEngine" endpoint="unix:///var/run/docker.sock" err="dial unix /var/run/docker.sock: connection refused"
2016-04-19T14:04:07Z [INFO] Beginning Polling for updates
2016-04-19T14:04:07Z [INFO] Initializing stats engine module="stats"
2016-04-19T14:04:07Z [INFO] Creating poll dialer module="ws client" host="ecs-a-2.us-east-1.amazonaws.com"
2016-04-19T14:04:07Z [INFO] Creating poll dialer module="ws client" host="ecs-t-2.us-east-1.amazonaws.com"
2016-04-19T14:04:15Z [INFO] Saving state! module="statemanager"
```

My service config:

```
rancher:
  services:
    ecs-agent-custom:
       image: amazon/amazon-ecs-agent
       labels:
          io.rancher.os.scope: system
          io.rancher.os.after: docker
       restart: always
       volumes:
         - /var/run:/var/run
         - /var/log/ecs/:/log
         - /var/lib/ecs/data:/data
         - /sys/fs/cgroup:/sys/fs/cgroup:ro
         - /var/run/docker/execdriver/native:/var/lib/docker/execdriver/native:ro
       ports:
         - "127.0.0.1:51678:51678"
       environment:
         - ECS_LOGFILE=/log/ecs-agent.log
         - ECS_LOGLEVEL=info
         - ECS_DATADIR=/data
         - ECS_*
         - AWS_*
```

This should definitely get changed for the ECS-optimized AMIs!

**Comment by schuylr** [comment]
Unfortunately there's a new issue by me doing this. It seems that a collection of attributes is created for the instance when it connects to the ECS cluster - and if the agent can't access the Docker socket on initialization, the agent connects with an empty set of attributes which don't get updated later. I can't initialize any tasks on the newly connected instances due to an `ATTRIBUTES` and `RESOURCE:CPU` error!

For now my only recourse is to run the `ecs-agent` service on the user docker instead of the system docker.

**Comment by joshwget** [comment]
@schuylr Yeah, what you described is what we currently suspect to be the issue.

When you run your custom service on the ECS-optimized AMI, is the error the same as with the regular service? Does it occur with the same frequency?

**Comment by schuylr** [comment]
@joshwget I don't use the ECS-optimized AMI when I use the custom service that I described above.

My custom service errors until the `docker.sock` file exists inside of `/var/run` - which you can see from my comment above. In the original scenario, the `ecs-agent.log` would just be full of the error mentioned in my first comment. And while this connected, it still didn't let me run tasks. The only way `ecs-agent` seems to work is by having it run with the user docker instead.

So, my latest config is this:

```
rancher:
  services:
    ecs-agent-custom:
       image: amazon/amazon-ecs-agent
       restart: always
       volumes:
         - /var/run/docker.sock:/var/run/docker.sock
         - /var/log/ecs/:/log
         - /var/lib/ecs/data:/data
         - /sys/fs/cgroup:/sys/fs/cgroup:ro
         - /var/run/docker/execdriver/native:/var/lib/docker/execdriver/native:ro
       ports:
         - "127.0.0.1:51678:51678"
       environment:
         - ECS_LOGFILE=/log/ecs-agent.log
         - ECS_LOGLEVEL=info
         - ECS_DATADIR=/data
         - ECS_*
         - AWS_*
```

**Comment by joshwget** [comment]
Oh I see. I thought you'd tried the custom service on both the optimized and non-optimized AMIs and that it only worked with the latter. Wait on the docker sock should be possible, so I think we'll still try to make this work with system docker. I'm glad running in user docker is working in the meantime though!

**Comment by schuylr** [comment]
@joshwget I think what's happening is that the system docker _is_ waiting for the user docker to start, but the user docker doesn't immediately set up `docker.sock` for `ecs-agent` to connect to when the user docker does start.

Does that make any sense?

**Comment by joshwget** [comment]
Right, I think waiting on `docker.sock` and not docker itself is what we'll try.

**Comment by ryansch** [comment]
For now you could install an entrypoint that runs something like the following:

```sh
if [ ! -S /var/run/docker.sock ]; then
  echo 'Docker socket not found, exiting.'
  exit 1
fi

# Guarantee that we can talk to docker or ecs-agent will start anyway
# but won't be able to do anything.
socat - UNIX-CONNECT:/var/run/docker.sock >/dev/null <<EOF
GET /info HTTP/1.1

EOF
if [ $? -ne 0 ]; then
  echo 'Failed to connect to /var/run/docker.sock, exiting.'
  exit 1
fi
```

**Comment by ryansch** [comment]
I've created a docker image with a working entrypoint: https://hub.docker.com/r/outstand/amazon-ecs-agent/

```yaml
ecs-agent:
  image: outstand/amazon-ecs-agent:wait
  labels:
    io.rancher.os.scope: system
    io.rancher.os.after: docker
  restart: always
  volumes:
    - /var/run/docker.sock:/var/run/docker.sock
    - /var/log/ecs/:/log
    - /var/lib/ecs/data:/data
    - /sys/fs/cgroup:/sys/fs/cgroup:ro
    - /var/run/docker/execdriver/native:/var/lib/docker/execdriver/native:ro
  ports:
    - "127.0.0.1:51678:51678"
  environment:
    - ECS_LOGFILE=/log/ecs-agent.log
    - ECS_LOGLEVEL=info
    - ECS_DATADIR=/data
    - ECS_*
    - AWS_*
```

**Comment by ryansch** [comment]
I have a more portable solution for consideration for v0.5.0.

Service:
```yaml
    wait-for-docker:
      image: rancher/os-docker:v0.4.5
      command: wait-for-docker
      labels:
        io.rancher.os.scope: system
        io.rancher.os.after: docker
        io.rancher.os.detach: "false"
      pid: host
      privileged: true
      volumes_from:
        - command-volumes
        - system-volumes
```

Surprisingly, `wait-for-docker` is already available in any container that mounts `command-volumes`.

**Comment by dr3s** [comment]
I still have this issue for 0.5.0 ECS optimized AMI.  Was there ever any fix applied?

**Comment by deniseschannon** [comment]
Using v0.6.0-rc5, I was able to use ECS Optimized instance to spin up ECS.
